iwyu_ast_util.cc

```diff
@@ -1299,12 +1325,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
     return retval;
 
   // Get the list of template args that apply to the decls.
-  const NamedDecl* decl = TypeToDeclForContent(tpl_spec_type);
-  const ClassTemplateSpecializationDecl* tpl_decl = DynCastFrom(decl);
-  if (!tpl_decl)   // probably because tpl_spec_type is a dependent type
-    return retval;
-  const TemplateArgumentList& tpl_args
-      = tpl_decl->getTemplateInstantiationArgs();
+  const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
+
+  const TemplateArgument* tpl_args = tpl_spec_type->getArgs();
+  unsigned num_args = tpl_spec_type->getNumArgs();
+
+  // When this is a ClassTemplateSpecializationDecl specifically, we want to
```

Comment:
And maybe move this comment up above the "default" case where we pull the args/num_args from the tpl_spec_type, to make it clearer that it all belongs together? Something like:
```
// Pull the template arguments out of the specialization type. If this is
// a ClassTemplateSpecializationDecl specifically, we want to
// get the arguments therefrom to correctly handle default arguments.
const TemplateArgument* tpl_args = tpl_spec_type->getArgs();
unsigned num_args = tpl_spec_type->getNumArgs();
const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
if (const ClassTemplateSpecializationDecl* cls_tpl_decl = DynCastFrom(decl)) {
  const TemplateArgumentList& tpl_arg_list = cls_tpl_decl->getTemplateInstantiationArgs();
  tpl_args = tpl_arg_list.data();
  num_args = tpl_arg_list.size();
}
```

or something like that.

Reply:
I find the loop below a little confusing when it's _not_ a `ClassTemplateSpecializationDecl` -- it seems like it's just running all types against themselves. But on the whole, I don't understand what this function does, so maybe that's something I just need to work on :)

Reply:
I don't really understand the function either, so this is something of the blind leading the blind...